src/languageHandler.js

```diff
@@ -64,75 +61,112 @@ export function _t(...args) {
+ * @param {object} variables Variable substitutions, e.g { foo: 'bar' }
+ * @param {object} tags Tag substitutions e.g. { 'a': (sub) => <a>{sub}</a> }
  *
- * @param {Function|Function[]} subs A function which will be called
- * with multiple args, each arg representing a captured group of the matching regexp.
- * This function must return a JSX node.
+ * The values to substitute with can be either simple strings, or functions that return the value to use in
+ * the substitution (e.g. return a React component). In case of a tag replacement, the function receives as
+ * the argument the text inside the element corresponding to the tag.
  *
- * @return a React <span> component containing the generated text
+ * @return a React <span> component if any non-strings were used in substitutions, otherwise a string
  */
-export function _tJsx(jsxText, patterns, subs) {
-    // convert everything to arrays
-    if (patterns instanceof RegExp) {
-        patterns = [patterns];
-    }
-    if (subs instanceof Function) {
-        subs = [subs];
-    }
-    // sanity checks
-    if (subs.length !== patterns.length || subs.length < 1) {
-        throw new Error(`_tJsx: programmer error. expected number of RegExps == number of Functions: ${subs.length} != ${patterns.length}`);
-    }
-    for (let i = 0; i < subs.length; i++) {
-        if (!(patterns[i] instanceof RegExp)) {
-            throw new Error(`_tJsx: programmer error. expected RegExp for text: ${jsxText}`);
+export function substitute(text, variables, tags) {
+    const regexpMapping = {};
+
+    if(variables !== undefined) {
+        for (const variable in variables) {
+            regexpMapping[`%\\(${variable}\\)s`] = variables[variable];
         }
-        if (!(subs[i] instanceof Function)) {
-            throw new Error(`_tJsx: programmer error. expected Function for text: ${jsxText}`);
+    }
+
+    if(tags !== undefined) {
+        for (const tag in tags) {
+            regexpMapping[`(<${tag}>(.*?)<\\/${tag}>|<${tag}>|<${tag}\\s*\\/>)`] = tags[tag];
         }
     }
+    return replaceByRegexes(text, regexpMapping);
+}
 
-    // The translation returns text so there's no XSS vector here (no unsafe HTML, no code execution)
-    const tJsxText = _t(jsxText, {interpolate: false});
-    const output = [tJsxText];
+/*
+ * Replace parts of a text using regular expressions
+ * @param {string} text The text on which to perform substitutions
+ * @param {object} mapping A mapping from regular expressions in string form to replacement string or a
+ * function which will receive as the argument the capture groups defined in the regexp. E.g.
+ * { 'Hello (.?) World': (sub) => sub.toUpperCase() }
+ *
+ * @return a React <span> component if any non-strings were used in substitutions, otherwise a string
+ */
+export function replaceByRegexes(text, mapping) {
+    const output = [text];
+
+    let wrap = false; // Remember if the output needs to be wrapped later
```

Comment:
I would go with what `wrap` _is_ as opposed to what will be done with `wrap`. So in this case, "// Whether the output should be wrapped by a <span> ...", explaining why this is being done as well

Reply:
A rename might help here, too. `shouldWrapInSpan` perhaps? (When I saw "wrap", I thought it referred to wrapping to the next line)